### PR TITLE
Job skill framework + medical scanner example

### PR DESCRIPTION
**Pull request by Hopekz**
Adds a job skill framework. 
This can be added to literally any action in the game including interacting with items or clicking a hud button.
increases RP

**What this means:**

This gives the ability to make any job speciality better at their actual job vs some random guy not in your department. You can use this framework to add time delays to actions to represent people stumbling around trying to figure something out or block them from performing an action outright.

As an example of this in action I have used this framework to add a medical check to the health scanner. This means that someone not trained in how to use medical equipment will have to wait 3 seconds to scan something and 1 second to adjust the settings but someone trained will be able to do it instantly as per usual before this PR.

**TRAINED PROFESSIONAL:**
O5woJ23yUa

**A random person without medical job skills:**
notice the progress bar over the medical scanner
b5f5ScveAW

:cl:  Hopek
rscadd: Added a Job skill framework. This gives the ability to make any job speciality better at their actual job vs some random guy not in your department. You can use this framework to add time delays to actions to represent people stumbling around trying to figure something out or block them from performing an action outright.
rscadd: Used this framework to add a medical check to the health scanner. This means that someone not trained in how to use medical equipment will have to wait 3 seconds to scan something and 1 second to adjust the settings but someone trained will be able to do it instantly as per usual before this PR.
/:cl:

### CODE REVIEW

**Comment by Marmio64** [comment]
make sure you add to the description that it increases rp or it won't be merged, hopek

**Comment by Hopekz** [comment]
> make sure you add to the description that it increases rp or it won't be merged, hopek

good catch I managed to change it before it was closed

File: code/__HELPERS/skills.dm [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+The actual skill check execution can be found under do_after() in mobs.dm
+skill checks can be called with do_after()
+
+Example: 
+do_after(user, skill = SKILL_MEDICAL, target = src) . This will require a basic medical check.
+
+
+Skills are something that will create a time block if the skill is not present. The duration of this is defined by skill_delay in do_after(). By default it is 3 seconds.
+You can set skill_required to TRUE in do_after() to reject an individual if they  do not meet requirements for the skill. This should probably be reserved for very specific tasks aimed specifically to stop cross department powergaming.
```
Comment (Hopekz) [inline code/__HELPERS/skills.dm]:
```suggestion
Skills are something that will create a time block if the skill is not present. The duration of this is defined by skill_delay in do_after(). By default, it is 3 seconds.
You can set skill_required to TRUE in do_after() to reject an individual if they do not meet requirements for the skill. This should probably be reserved for very specific tasks aimed specifically to stop cross-department power gaming.
```

File: code/modules/jobs/job_types/ai.dm [hunk 2]
```diff
@@ -15,6 +15,7 @@
 	exp_type_department = EXP_TYPE_SILICON
 	display_order = JOB_DISPLAY_ORDER_AI
 	var/do_special_check = TRUE
+	job_skills = list(SKILL_DEBUG) // silicons and AI get debug to bypass skill checks until there is a need to give AI or cyborg specific skills
```
Comment (ynot01) [inline code/modules/jobs/job_types/ai.dm:18]: i think it would be great to have different skills loaded into individual modules

**Comment by alexkar598** [comment]
Three things:
1) This.... already kinda exists using traits, such as the surgery trait that allows doctors to perform surgery with a better success rate

2) The health analyzer is a bad candidate for this, atleast for a delay, when you need to give emergency medical attention to someone, 3 seconds is alot of time to waste on scanning person(nearly impossible to get someone to stand still that long either), it's also not that uncommon to spam the health analyzer to get an updating feed.

3) This doesnt really address the core issue of medbay being self serve at all, while non med people healing other people does occur, it's not a huge issue, the biggest issue being people able to heal themselves easily, which they can still do by correlating the health doll to the examine menu.

Also it doesn't really matter but this may confuse players as they dont know that the health analyzer is slow because of their job

Tl;dr
-  system already kinda exists and having 2 systems isnt ideal
-  health analyzer is a bad candidate to see skill gating atleast currently
-  If you add a delay on the health analyzer, you might as well make it not usable

**Comment by Hopekz** [comment]
> Tl;dr
> 
> * system already kinda exists and having 2 systems isnt ideal

While this does in a sense exist this framework makes it very easy to integrate it into tasks and it is already pre-separated.
People have had traits for a while but they are inconvenient to work with in terms of making these types of changes.

Restricting something to security can be done with a simple ' do_after(user, skill = SKILL_SECURITY, skill_required = TRUE)  ' without thinking about it too hard. 
You don't have to worry about security having this "trait" you just know all of them get it when they get their jobs.

**Comment by Hopekz** [comment]
> * health analyzer is a bad candidate to see skill gating atleast currently
> * If you add a delay on the health analyzer, you might as well make it not usable

**I separated this reply to let people vote up and down on this.** 

I'm fine with not adding any examples to the framework PR I just thought this is something that would work well since I've seen it on CM and really liked it as a feature.

Will update to either remove that example (down doinks) or keep it (up doinks) in about a day or two

**Comment by Marmio64** [comment]
> > * health analyzer is a bad candidate to see skill gating atleast currently
> > * If you add a delay on the health analyzer, you might as well make it not usable
> 
> **I separated this reply to let people vote up and down on this.**
> 
> I'm fine with not adding any examples to the framework PR I just thought this is something that would work well since I've seen it on CM and really liked it as a feature.
> 
> Will update to either remove that example (down doinks) or keep it (up doinks) in about a day or two

health analyzers (at least according to charliestation lore) are supposed to be usable by non medically trained people but they're probably still quite clunky and might need some knowledge on it to work. Health analyzers aren't really needed for stabilizing people since you can give examine them to see if they're bruised or something, apply a patch, epipen, and carry to medbay rather than healing them yourself. However, maybe advanced analyzers could maybe be usable instantly by everyone since they're more streamlined even if they have more features. This would also allow medbay to still recruit new people who are just as good since they'd be using advanced tools that supplement their lack of knowledge, even if it takes a while to research it first.

File: code/__HELPERS/mobs.dm [hunk 3]
```diff
@@ -1,3 +1,5 @@
+#define HAS_SKILL(target, skill) (target.mind.skills ? (target.mind.skills[skill] ? TRUE : FALSE) : FALSE)
```
Comment (alexkar598) [inline code/__HELPERS/mobs.dm:1]: Unneeded ternary operator

File: code/__HELPERS/mobs.dm [hunk 4]
```diff
@@ -247,9 +249,18 @@ GLOBAL_LIST_EMPTY(species_list)
 		checked_health["health"] = health
 	return ..()
 
-/proc/do_after(mob/user, var/delay, needhand = 1, atom/target = null, progress = 1, datum/callback/extra_checks = null, stayStill = TRUE)
+/proc/do_after(mob/user, var/delay = 0, needhand = 1, atom/target = null, progress = 1, datum/callback/extra_checks = null, stayStill = TRUE, var/skill = "undefined", skill_required = FALSE, skill_delay = 30)
 	if(!user)
 		return 0
+	if(skill != "undefined") // skills are required here
```
Comment (alexkar598) [inline code/__HELPERS/mobs.dm:255]: Should default to null, not a string

File: code/__HELPERS/mobs.dm [hunk 5]
```diff
@@ -247,9 +249,18 @@ GLOBAL_LIST_EMPTY(species_list)
...
+		if(!user.mind?.skills.Find(skill) && !user.mind?.skills.Find("divine"))
```
Comment (alexkar598) [inline code/__HELPERS/mobs.dm:256]: Needs to use HAS_SKILL, also shouldnt be using a string instead of a define for the second check

File: code/__HELPERS/mobs.dm [hunk 6]
```diff
@@ -247,9 +249,18 @@ GLOBAL_LIST_EMPTY(species_list)
...
 		return 0
+	if(skill != "undefined") // skills are required here
+		if(!user.mind?.skills.Find(skill) && !user.mind?.skills.Find("divine")) 
+			if(skill_required)
+				to_chat(user,"<span class='warning'> You need advanced [skill] skill in order to perform this task.")
+				return 0
+			else
+				to_chat(user,"<span class='notice'> You struggle as you do not have [skill] skill.")
+				delay = delay + skill_delay
+
```
Comment (alexkar598) [inline code/__HELPERS/mobs.dm:263]: Check if delay is 0 and early return true if it is

File: code/__HELPERS/skills.dm [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+Example: 
+do_after(user, skill = SKILL_MEDICAL, target = src) . This will require a basic medical check.
+
+
+Skills are something that will create a time block if the skill is not present. The duration of this is defined by skill_delay in do_after(). By default, it is 3 seconds.
+You can set skill_required to TRUE in do_after() to reject an individual if they do not meet requirements for the skill. This should probably be reserved for very specific tasks aimed specifically to stop cross-department power gaming.
+*/
+
+
+#define SKILL_PLACEHOLDER "undefined" // here for debug + placeholder for do_after() proc
```
Comment (alexkar598) [inline code/__HELPERS/skills.dm:16]: This is unneeded, use null

File: code/__HELPERS/mobs.dm [hunk 8]
```diff
@@ -247,9 +249,18 @@ GLOBAL_LIST_EMPTY(species_list)
 		checked_health["health"] = health
 	return ..()
 
-/proc/do_after(mob/user, var/delay, needhand = 1, atom/target = null, progress = 1, datum/callback/extra_checks = null, stayStill = TRUE)
+/proc/do_after(mob/user, var/delay = 0, needhand = 1, atom/target = null, progress = 1, datum/callback/extra_checks = null, stayStill = TRUE, var/skill = "undefined", skill_required = FALSE, skill_delay = 30)
```
Comment (alexkar598) [inline code/__HELPERS/mobs.dm:252]: Dont have a default skill delay of 3 seconds, thats weird and should always be defined when you call the proc, not by default

File: code/modules/jobs/job_types/_job.dm [hunk 9]
```diff
@@ -35,6 +35,9 @@
 	//Sellection screen color
 	var/selection_color = "#ffffff"
 
+	// skills people with this job have
+	var/list/job_skills = list(SKILL_PLACEHOLDER)
```
Comment (alexkar598) [inline code/modules/jobs/job_types/_job.dm:39]: Why does this need to exist instead of being an empty list

File: code/modules/jobs/job_types/chief_engineer.dm [hunk 10]
```diff
@@ -15,17 +15,18 @@
 	exp_requirements = 180
 	exp_type = EXP_TYPE_CREW
 	exp_type_department = EXP_TYPE_ENGINEERING
+	job_skills = list(SKILL_COMMAND, SKILL_ENGINEERING, SKILL_ATMOS, SKILL_TCOMS)
```
Comment (alexkar598) [inline code/modules/jobs/job_types/chief_engineer.dm:18]: Setting SKILL_COMMAND, SKILL_ENGINEERING and friends on each job type is rather error prone if someone adds a job or a new skill later on, why not set it dynamically based on the jobs in this file

https://github.com/yogstation13/Yogstation/blob/8982248053ac9d2a04f0d8172004fe4b2e2703c2/code/modules/jobs/jobs.dm

File: code/modules/jobs/job_types/research_director.dm [hunk 11]
```diff
@@ -15,19 +15,20 @@
...
+						ACCESS_TOX_STORAGE, ACCESS_TELEPORTER, ACCESS_SEC_DOORS, ACCESS_MECH_SCIENCE,
+						ACCESS_RESEARCH, ACCESS_ROBOTICS, ACCESS_XENOBIOLOGY, ACCESS_AI_UPLOAD,
+						ACCESS_RC_ANNOUNCE, ACCESS_KEYCARD_AUTH, ACCESS_GATEWAY, ACCESS_MINERAL_STOREROOM,
+						ACCESS_TECH_STORAGE, ACCESS_MINISAT, ACCESS_MAINT_TUNNELS, ACCESS_NETWORK)
 	minimal_access = list(ACCESS_RD, ACCESS_HEADS, ACCESS_TOX, ACCESS_GENETICS, ACCESS_MORGUE,
-			            ACCESS_TOX_STORAGE, ACCESS_TELEPORTER, ACCESS_SEC_DOORS, ACCESS_MECH_SCIENCE,
-			            ACCESS_RESEARCH, ACCESS_ROBOTICS, ACCESS_XENOBIOLOGY, ACCESS_AI_UPLOAD,
-			            ACCESS_RC_ANNOUNCE, ACCESS_KEYCARD_AUTH, ACCESS_GATEWAY, ACCESS_MINERAL_STOREROOM,
-			            ACCESS_TECH_STORAGE, ACCESS_MINISAT, ACCESS_MAINT_TUNNELS, ACCESS_NETWORK)
+						ACCESS_TOX_STORAGE, ACCESS_TELEPORTER, ACCESS_SEC_DOORS, ACCESS_MECH_SCIENCE,
```
Comment (alexkar598) [inline code/modules/jobs/job_types/research_director.dm:28]: Whyyyyyyyyyyyyyy?

**Review by alexkar598 — CHANGES_REQUESTED** [review]

**Review by alexkar598 — CHANGES_REQUESTED** [review]
This is simply too complicated, use mind.job

if(mind.job == "Janitor" || do_after(user, 30, target=src))
